### PR TITLE
Repository transfer has to be confirmed, if user can not create repo for new owner

### LINKED ISSUE #4743: Transfer repo ownership should be confirmed by new owner
- Gitea version (or commit ref): b1ad573 
- Git version: any
- Operating system: any
- Database (use `[x]`): any
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) https://try.gitea.io/test/frame-test
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

User can easily frame another user. How to reproduce:
1. Create new repo.
2. Configure local git to use victim's name/email.
3. Add to this repo something illegal in victim's country (or something agressive about other group of people or just smething very shameful).
4. Transfer ownership of repo to victim.
5. Call the police (or press, victim's boss, victim's wife or anyone else who will hate victim for publishing such information).

From (2) there is no protection, because it's how git itself works (it allows unsigned commits by design).
But from (4) there should be some kind of protection. For example, on Github anyone cannot simply transfer repo ownership to anyone, new owner should confirm this transfer.

Between user/org, user should be able to transfer repo without confirmation if he has admin rights in this org. If transferring repo to another org, any org admin should be able to confirm this transfer. If transferring repo to another user, another user should confirm transfer. And actual transfer must be performed after confirmation. And if user has admin rights, he probably should be able to transfer repo without confirmation (as he implicitly has all privileges on server).

Upd: additionally, when transfer is pending and waiting for user confirmation, new repo owner should have temporary read access to repo if it's private - to make him able to investigate this repo content before accepting/rejecting transfer.

## Screenshot

How it looks on Github:

### PR DESCRIPTION
**Pull request by 6543**
[Screenshot at 2021-02-26 21-53-47](https://cloud.obermui.de/s/moXx26RgLDSxM6d)

close #4743

### CODE REVIEW

File: services/mailer/mail_repo.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+// SendRepoTransferNotifyMail triggers a notification e-mail when a repository
+// transfer is initiated
+func SendRepoTransferNotifyMail(locale Locale, u *models.User, repo *models.Repository) {
+	data := map[string]interface{}{
+		"Subject":             locale.Tr("mail.repo_transfer_notify"),
+		"RepoName":            repo.FullName(),
+		"Link":                repo.HTMLURL(),
+		"AcceptTransferLink":  repo.HTMLURL() + "/action/accept_transfer",
```
Comment (6543) [inline services/mailer/mail_repo.go:22]: new routes needed!

File: routers/repo/setting.go [hunk 2]
```diff
@@ -477,18 +477,59 @@ func SettingsPost(ctx *context.Context) {
...
+				ctx.RenderWithErr(ctx.Tr("repo.settings.transfer_in_progress"), tplSettingsOptions, nil)
 			} else {
 				ctx.ServerError("TransferOwnership", err)
 			}
+
+			return
+		}
+
+		if setting.Service.EnableNotifyMail {
+			mailer.SendRepoTransferNotifyMail(ctx.Locale, ctx.Repo.Owner, ctx.Repo.Repository)
```
Comment (6543) [inline routers/repo/setting.go]: mv to repo_service

File: routers/repo/setting.go [hunk 3]
```diff
@@ -477,18 +477,59 @@ func SettingsPost(ctx *context.Context) {
...
+		}
+
+		log.Trace("Repository transfer process was started: %s/%s -> %s", ctx.Repo.Owner.Name, repo.Name, newOwner)
+		ctx.Flash.Success(ctx.Tr("repo.settings.transfer_started", newOwner.LowerName))
+		ctx.Redirect(setting.AppSubURL + "/" + ctx.Repo.Owner.Name + "/" + repo.Name + "/settings")
+
+	case "cancel_transfer":
```
Comment (6543) [inline routers/repo/setting.go:501]: own files & routes

-> https://github.com/go-gitea/gitea/pull/14792/files#r582535386

File: models/repo_transfer.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	if _, err = sess.Insert(transfer); err != nil {
+		return err
+	}
+
+	// TODO: user witch have "create repo"-access to new place
+
+	return sess.Commit()
+}
+
+// TransferOwnership transfers all corresponding setting from old user to new one.
```
Comment (6543) [inline models/repo_transfer.go]:
```suggestion
// TransferOwnership transfers all corresponding repository items from old user to new one.
```

File: routers/repo/repo.go [hunk 5]
```diff
@@ -293,6 +298,36 @@ func Action(ctx *context.Context) {
...
+	if err = repoTransfer.LoadAttributes(); err != nil {
+		return err
+	}
+
+	if !repoTransfer.CanUserAcceptTransfer(ctx.User) {
+		return errors.New("user does not have enough permissions")
+	}
+
+	if accept {
+		if err = repo_service.TransferOwnership(repoTransfer.Doer, repoTransfer.Recipient, ctx.Repo.Repository, repoTransfer.Teams); err != nil {
```
Comment (6543) [inline routers/repo/repo.go:316]: rm pending entry!

File: models/migrations/v174.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline models/migrations/v174.go]:
```suggestion
// Copyright 2021 The Gitea Authors. All rights reserved.
```

File: services/mailer/mail_repo.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline services/mailer/mail_repo.go]:
```suggestion
// Copyright 2021 The Gitea Authors. All rights reserved.
```

File: models/repo_transfer.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+		return ErrRepoAlreadyExist{newOwner.LowerName, repo.Name}
+	}
+
+	transfer := &RepoTransfer{
+		RepoID:      repo.ID,
+		RecipientID: newOwner.ID,
+		CreatedUnix: timeutil.TimeStampNow(),
+		UpdatedUnix: timeutil.TimeStampNow(),
+		DoerID:      doer.ID,
+		TeamIDs:     []int64{},
```
Comment (zeripath) [inline models/repo_transfer.go]:
```suggestion
		TeamIDs:     make([]int64, 0, len(teams)),
```

File: models/repo_transfer.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+	// If there was previously a redirect at this location, remove it.
+	if err = deleteRepoRedirect(sess, newOwner.ID, repo.Name); err != nil {
+		return fmt.Errorf("delete repo redirect: %v", err)
+	}
+
+	if err := newRepoRedirect(sess, oldOwner.ID, repo.ID, repo.Name, repo.Name); err != nil {
+		return fmt.Errorf("newRepoRedirect: %v", err)
+	}
+
+	return sess.Commit()
```
Comment (zeripath) [inline models/repo_transfer.go:334]: If the transaction fails we will be left repository in the wrong place

File: modules/convert/notification.go [hunk 10]
```diff
@@ -48,12 +48,18 @@ func ToNotificationThread(n *models.Notification) *api.NotificationThread {
 				result.Subject.LatestCommentURL = comment.APIURL()
 			}
 		}
-	case models.NotificationSourceCommit:
+	case models.NotificationSourceCommit: //unused until now
```
Comment (zeripath) [inline modules/convert/notification.go]: Don't think this comment is helpful or enlightening

```suggestion
	case models.NotificationSourceCommit:
```

File: services/mailer/mail_repo.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+		for i := range users {
+			emails = append(emails, users[i].Email)
+		}
+		destination = newOwner.DisplayName()
+	} else {
+		emails = []string{newOwner.Email}
+		destination = "you"
+	}
+
+	subject := fmt.Sprintf("%s like to transfer \"%s\" to %s", doer.DisplayName(), repo.FullName(), destination)
```
Comment (zeripath) [inline services/mailer/mail_repo.go]: Isn't there a way of localising this?

The message should be:

```suggestion
	subject := fmt.Sprintf("%s would like to transfer \"%s\" to %s", doer.DisplayName(), repo.FullName(), destination)
```

File: services/repository/transfer.go [hunk 12]
```diff
@@ -41,7 +41,7 @@ func TransferOwnership(doer, newOwner *models.User, repo *models.Repository, tea
 	}
 
 	for _, team := range teams {
-		if err := team.AddRepository(newRepo); err != nil {
+		if err = team.AddRepository(newRepo); err != nil {
```
Comment (zeripath) [inline services/repository/transfer.go]: Put the assignment back in - you're hurting the compiler's ability to use a throwaway assignment here.

```suggestion
		if err := team.AddRepository(newRepo); err != nil {
```

File: templates/repo/header.tmpl [hunk 13]
```diff
@@ -42,6 +42,24 @@
...
+						</div>
+					</form>
+					<form method="post" action="{{$.RepoLink}}/action/reject_transfer?redirect_to={{$.RepoLink}}">
+						{{$.CsrfTokenHtml}}
+						<div class="ui poping up" data-content="{{if $.CanUserAcceptTransfer}}{{$.i18n.Tr "repo.transfer.reject_desc" $.RepoTransfer.Recipient.DisplayName}}{{else}}{{$.i18n.Tr "repo.transfer.no_permission_to_reject"}}{{end}}" data-position="bottom center" data-variation="tiny">
+							<button type="submit" class="ui button {{if $.CanUserAcceptTransfer}}red {{end}}ok inverted small"{{if not $.CanUserAcceptTransfer}} disabled{{end}}>
+								{{$.i18n.Tr "repo.transfer.reject"}}
+							</button>
+						</div>
+					</form>
```
Comment (zeripath) [inline templates/repo/header.tmpl]: indenting...

File: templates/repo/settings/options.tmpl [hunk 14]
```diff
@@ -444,11 +444,23 @@
...
+						<button class="ui basic red show-modal button" data-modal="#transfer-repo-modal">{{.i18n.Tr "repo.settings.transfer"}}</button>
+					{{ end }}
 				</div>
 				<div>
 					<h5>{{.i18n.Tr "repo.settings.transfer"}}</h5>
+					{{if .RepoTransfer}}
+					<p>{{.i18n.Tr "repo.settings.transfer_started" .RepoTransfer.Recipient.DisplayName}}</p>
+				        {{else}}
 					<p>{{.i18n.Tr "repo.settings.transfer_desc"}}</p>
+					{{end}}
```
Comment (zeripath) [inline templates/repo/settings/options.tmpl]: Indenting

File: models/repo_transfer.go (hunk 9 above)
Comment (6543) [reply models/repo_transfer.go:334]: created an issue, as this is code from master & worth it's own pull
-> https://github.com/go-gitea/gitea/issues/14821

File: services/mailer/mail_repo.go (hunk 11 above)
Comment (6543) [reply services/mailer/mail_repo.go]: > ... a way of localising

~~not sure I did stick with the eng one as this is most similar with how other email are generated - and one thing we should have in mind to improve ... :/~~

no

Comment (6543) [reply services/mailer/mail_repo.go]: created https://github.com/go-gitea/gitea/issues/14822

Comment (zeripath) [reply services/mailer/mail_repo.go]: https://github.com/go-gitea/gitea/blob/3d8b5ad5f37f495e50d83bbad214f8061f8a5ac4/services/mailer/mail.go#L227-L232

**Review by zeripath — APPROVED** [review]

File: models/repo_transfer.go [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+		return ErrRepoTransferInProgress{}
+	}
+	return nil
+}
+
+// CreatePendingRepositoryTransfer transfer a repo from one owner to a new one.
+// it marks the repository transfer as "pending"
+func CreatePendingRepositoryTransfer(doer, newOwner *User, repoID int64, teams []*Team) error {
+	sess := x.NewSession()
+	defer sess.Close()
```
Comment (lafriks) [inline models/repo_transfer.go:150]: This should be moved after sess.Begin

File: models/repo_transfer.go [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// TransferOwnership transfers all corresponding repository items from old user to new one.
+func TransferOwnership(doer *User, newOwnerName string, repo *Repository) error {
+	sess := x.NewSession()
+	defer sess.Close()
```
Comment (lafriks) [inline models/repo_transfer.go:200]: Same here

File: models/repo_transfer.go (hunk 15 above)
Comment (zeripath) [reply models/repo_transfer.go:150]: hmm... we always need to close a session whether it has begun or not:

https://github.com/go-gitea/gitea/blob/dc081959dbf5aea99f975b79275d43efe681bc25/models/avatar.go#L92-L98

etc.

File: models/repo_transfer.go [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+// CancelRepositoryTransfer marks the repository as ready and remove pending transfer entry,
+// thus cancel the transfer process.
+func CancelRepositoryTransfer(repo *Repository) error {
+	sess := x.NewSession()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	defer sess.Close()
```
Comment (lafriks) [inline models/repo_transfer.go]: Here

Comment (6543) [reply models/repo_transfer.go]: jup!

**Review by lafriks — APPROVED** [review]

**Pull request merged**